omniledger/service/proof.go

```diff
@@ -0,0 +1,106 @@
+package service
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/dedis/student_18_omniledger/omniledger/collection"
+	"gopkg.in/dedis/cothority.v2"
+	"gopkg.in/dedis/cothority.v2/skipchain"
+	"gopkg.in/dedis/onet.v2/network"
+)
+
+// Proof represents everything necessary to verify a given
+// key/value pair is stored in a skipchain. The proof is in three parts:
+//   1. Collection proofs the presence or absence of the key. In case of
+//   the key being present, the value is included in the proof
+//   2. Latest is used to verify the merkle tree root used in the collection-proof
+//   is stored in the latest skipblock
+//   3. Links proves that the latest skipblock is part of the skipchain
+//
+// This Structure could later be moved to cothority/skipchain.
+type Proof struct {
+	// Proving that the key/value pair is in the collection.
+	Collection collection.Proof
+	// Providing the latest skipblock to retrieve the Merkle tree root.
+	Latest skipchain.SkipBlock
+	// Proving the path to the latest skipblock.
+	Links []skipchain.ForwardLink
+}
+
+// NewProof creates a proof for key in the skipchain with the given id. It uses
+// the collectionDB to look up the key and the skipblockdb to create the correct
+// proof for the forward links.
+func NewProof(c *collectionDB, s *skipchain.SkipBlockDB, id skipchain.SkipBlockID,
+	key []byte) (p *Proof, err error) {
+	p = &Proof{}
+	p.Collection, err = c.coll.Get(key).Proof()
+	if err != nil {
+		return
+	}
+	sb := s.GetByID(id)
+	if sb == nil {
+		return nil, errors.New("didn't find skipchain")
+	}
+	for len(sb.ForwardLink) > 0 {
+		link := sb.ForwardLink[len(sb.ForwardLink)-1]
+		p.Links = append(p.Links, *link)
+		sb = s.GetByID(link.To)
+		if sb == nil {
+			return nil, errors.New("missing block in chain")
+		}
+	}
+	p.Latest = *sb
+	return
+}
+
+// ErrorVerifyCollection is returned if the collection-proof itself
+// is not properly set up.
+var ErrorVerifyCollection = errors.New("collection inclusion proof is wrong")
+
+// ErrorVerifyMerkleRoot is returned if the root of the collection
+// is different than the stored value in the skipblock.
+var ErrorVerifyMerkleRoot = errors.New("root of collection is not in skipblock")
+
+// ErrorVerifySkipchain is returned if the stored skipblock doesn't
+// have a proper proof that it comes from the genesis block.
+var ErrorVerifySkipchain = errors.New("stored skipblock is not properly evolved from genesis block")
+
+// Verify takes a skipchain id and verifies that the proof is valid for this skipchain.
+// It verifies the collection-proof, that the merkle-root is stored in the skipblock
+// of the proof and the fact that the skipblock is indeed part of the skipchain.
+// If all verifications are correct, the error will be nil.
+func (p Proof) Verify(genesis *skipchain.SkipBlock) error {
```

Comment:
Why doesn't Verify do p.Match? I feel it should be a part of this function, the user might forget to call match and think Verify is all that's needed.

Reply:
We can have a proof that the key is in there or a proof that the key is missing. A transaction will be one of `Create`, `Update`, `Delete`. So in the case of `Delete` you want to be sure that `Verify` is OK, but that `Match` returns 'false'.